src/Microsoft.AspNetCore.OData/Formatter/Serialization/ODataResourceSetSerializer.cs

```diff
@@ -120,6 +131,73 @@ await WriteObjectInlineAsync(graph, resourceSetType, writer, writeContext)
 
             Func<object, Uri> nextLinkGenerator = GetNextLinkGenerator(resourceSet, enumerable, writeContext);
 
+            WriteResourceSetInternal(resourceSet, elementType, resourceSetType, writeContext, out bool isUntypedCollection, out IODataEdmTypeSerializer resourceSerializer);
+
+            await writer.WriteStartAsync(resourceSet).ConfigureAwait(false);
+            object lastResource = null;
+
+            foreach (object item in enumerable)
+            {
+                lastResource = item;
+
+                await WriteResourceSetItemAsync(item, elementType, isUntypedCollection, resourceSetType, writer, resourceSerializer, writeContext);
+            }
+
+            // Subtle and surprising behavior: If the NextPageLink property is set before calling WriteStart(resourceSet),
+            // the next page link will be written early in a manner not compatible with odata.streaming=true. Instead, if
+            // the next page link is not set when calling WriteStart(resourceSet) but is instead set later on that resourceSet
+            // object before calling WriteEnd(), the next page link will be written at the end, as required for
+            // odata.streaming=true support.
+
+            resourceSet.NextPageLink = nextLinkGenerator(lastResource);
+
+            await writer.WriteEndAsync().ConfigureAwait(false);
+        }
+
+        private async Task WriteResourceSetAsync(IAsyncEnumerable<object> asyncEnumerable, IEdmTypeReference resourceSetType, ODataWriter writer,
+           ODataSerializerContext writeContext)
+        {
+            Contract.Assert(writer != null);
+            Contract.Assert(writeContext != null);
+            Contract.Assert(asyncEnumerable != null);
+            Contract.Assert(resourceSetType != null);
+
+            IEdmStructuredTypeReference elementType = GetResourceType(resourceSetType);
+            ODataResourceSet resourceSet = CreateResourceSet(asyncEnumerable, resourceSetType.AsCollection(), writeContext);
+
+            Func<object, Uri> nextLinkGenerator = GetNextLinkGenerator(resourceSet, asyncEnumerable, writeContext);
+
+            WriteResourceSetInternal(resourceSet, elementType, resourceSetType, writeContext, out bool isUntypedCollection, out IODataEdmTypeSerializer resourceSerializer);
+            
+            await writer.WriteStartAsync(resourceSet).ConfigureAwait(false);
+            object lastResource = null;
+
+            await foreach (object item in asyncEnumerable)
+            {
+                lastResource = item;
+
+                await WriteResourceSetItemAsync(item, elementType, isUntypedCollection, resourceSetType, writer, resourceSerializer, writeContext);
```

Comment:
Should we call `ConfigureAwait(false)` here?

```suggestion
                await WriteResourceSetItemAsync(item, elementType, isUntypedCollection, resourceSetType, writer, resourceSerializer, writeContext).ConfigureAwait(false);
```